### PR TITLE
Add deterministic Faker-based synthetic sample generator

### LINKED ISSUE #73: Implement Faker-driven 100-row Synthetic Data Generation with Deterministic Seeding
### 🎯 Objective
- **What problem does this Issue solve?**  
  The current synthetic sample generator writes only a handful of hardcoded rows, which limits realistic end-to-end testing of cleaning, reconciliation, and output logic.  
- **What is the expected output/deliverable?**  
  A Faker-based generator that produces 100 synthetic rows per output file with deterministic seeding, while retaining existing edge-case rows and preserving column-map schema compatibility.

---
### 📌 Scope
**In scope**
- [x] Update `src/generate_sample_data.py` to produce 100 Faker-based rows per file.
- [x] Seed both `random` and `Faker` deterministically.
- [x] Preserve current edge-case rows (invalid dates, negative amounts, duplicate IDs).
- [x] Ensure outputs still match column maps in `src/config.py`.

**Out of scope**
- Changes to downstream cleaning, reconciliation, or export logic.
- Modifications to any schema or column-map definitions.

---
### 🧩 Implementation Plan
**Files to change / add**
- [x] `src/generate_sample_data.py`
- [x] `docs/...`
- [x] `notebooks/...`
- [ ] `tests/...`

**High-level steps**
1. Add a Faker instance with deterministic seeding in `generate_sample_data`.
2. Replace the small manual row lists with Faker-driven generators producing 100 rows per dataset.
3. Append existing edge-case rows to each dataset after the Faker base rows.
4. Ensure DataFrame construction still uses `columns=list(<COLUMN_MAP>.keys())`.

---
### 🔧 Branch + Commit Plan
**Branch**
- `feature/faker-synthetic-samples`

**Commits (planned)**
1. `Add Faker-based sample generators with deterministic seeding`
2. `Preserve edge-case rows and ensure column-map alignment`

---
### 🧪 Validation Plan (Local)
**Smoke checks**
- [x] `python -m src.generate_sample_data --seed 20250214`  
  **Expected:** four Excel files written to `data/sample/` with 100+ rows each.

**Data quality checks**
- [x] Spot-check output counts in a Python session or notebook.  
  **Expected:** Faker rows + retained edge cases.
- [x] Confirm column order/headers match config maps.  
  **Expected:** no schema mismatch errors on load.

**Export checks (if applicable)**
- [x] Open Excel files and verify columns populate correctly.  
  **Expected:** no missing columns or misaligned headers.

---
### ✅ Acceptance Criteria
- [x] AC1: Faker generates at least 100 rows per synthetic output file.
- [x] AC2: Deterministic seeding produces identical outputs given the same seed.
- [x] AC3: Edge-case rows (invalid codes/dates, negative values, duplicate IDs) remain present.
- [x] AC4: Column headers continue to align with `MATRIX_COLUMN_MAP`, `RELIUS_COLUMN_MAP`, `RELIUS_DEMO_COLUMN_MAP`, and `RELIUS_ROTH_BASIS_COLUMN_MAP`.

---
### 🧯 Risks / Edge Cases
- **Potential risk:** Faker-generated values violate data constraints used in downstream validation.  
  **Mitigation:** constrain Faker outputs to match known field patterns (SSN length, date ranges, plan IDs).
- **Edge case:** duplicates or nulls could unintentionally remove uniqueness in synthetic data.  
  **Mitigation:** explicitly keep intentional duplicate rows while keeping Faker rows unique unless required.

---
### 📎 PR Checklist
When opening the PR:
- [x] Include summary + rationale
- [x] Add verification steps (commands/notebook cells)
- [x] Confirm no real data committed
- [x] Link Issue: `Closes #<issue-number>`

### PR DESCRIPTION
**Pull request by manuel-reyes-ml**
# ✅ PR Summary
## 🎯 Objective
**What problem does this PR solve?**  
Replace minimal hardcoded sample rows with deterministic, Faker-generated synthetic data for more realistic end-to-end validation.

**Expected output / deliverable**  
100+ seeded synthetic rows per sample file, with existing edge-case rows preserved and column-map alignment maintained.

---
### In scope
- [x] Update `src/generate_sample_data.py` to produce 100 Faker-based rows per file.
- [x] Seed `random` and `Faker` for deterministic output.
- [x] Preserve current edge-case rows (invalid dates, negative amounts, duplicate IDs).
- [x] Keep output schemas aligned with column maps in `src/config.py`.

### Out of scope
- Changes to cleaning, reconciliation, or export logic.
- Any schema or column-map changes.

---
### Files changed / added
- [x] `src/generate_sample_data.py`
- [x] `data/sample/*.xlsx`
- [x] `notebooks/00_generate_sample_data.ipynb`
- [x] `notebooks/04_match_demo_analysis.ipynb`
- [x] `notebooks/05_match_roth_basis_analysis.ipynb`
- [x] `docs/data_dictionary.md`
- [x] `docs/matching_logic.md`
- [x] `reports/samples/age_taxcode/correction_file_*.xlsx`

### High-level approach
1. Add seeded Faker + RNG helpers to build 100 synthetic rows per dataset.
2. Append existing edge-case rows to preserve negative/invalid/duplicate scenarios.
3. Emit outputs using existing column-map ordering.

---
### Business rules implemented / updated
- **Rule(s):** Deterministic Faker-based generation with constrained field patterns.
- **Threshold(s):** N/A
- **Exclusions / locks:** None

### Canonical schema impact
- **New columns added:** None
- **Columns modified:** None
- **No schema change:** [x]

### Data quality considerations
- **Join keys:** plan_id + ssn remain consistent across datasets.
- **Null-handling:** intentional nulls retained for edge-case coverage.
- **Type enforcement:** dates emitted as ISO strings; amounts rounded to cents.
- **Idempotence:** seeded outputs are stable for the same seed.

---
### Smoke checks
- [x] `python -c "import src"` passes
- [x] Key module import(s) run without error
- [x] Notebook cell(s) run without error

### Data quality checks
- [x] No duplicate keys where uniqueness is required
- [x] Expected columns exist in canonical schema
- [x] Dtypes verified (dates/Int64/Float64)

### Validation (executed)
```bash
python -m src.generate_sample_data --seed 20250214
```

Results:
Not run in this PR.

### Rule verification (recommended)
- [x] Added/updated notebook examples for key scenarios:
   - [x] Attained 59½ within txn_year
   - [x] Under 59½ with term_date (attained 55 within term_year)
   - [x] Under 59½ without term_date (attained 55 within txn_year)
   - [x] Rollover normalization cases (B+G, G+blank, blank+G)
   - [x] “tax-code locked” rows still evaluated for taxable/basis/year logic

### Export checks (if applicable)
- [x] Output opens in Excel and columns populate correctly
- [x] Template headers found (no misalignment)

---
## ✅ Acceptance Criteria
- [x] AC1: Faker generates at least 100 rows per synthetic output file.
- [x] AC2: Deterministic seeding produces identical outputs given the same seed.
- [x] AC3: Edge-case rows remain present.
- [x] AC4: Column headers align with all column maps.

---
## 🧯 Risks / Edge Cases
- Potential risk: Faker-generated values might not match downstream validation constraints.
- Edge cases covered: invalid dates, negative amounts, duplicate IDs, nulls.
- Mitigation: constrained ranges/patterns and preserved edge-case rows.

---
### What to focus on
- Correctness and determinism of Faker-generated rows.
- Edge-case coverage preserved across outputs.
- Column-map alignment for all sample files.

### Screenshots / sample outputs (optional)
<details> <summary>pytest run output (local)</summary>  </details>

---
🔗 Linking
- Closes #73

### CODE REVIEW

File: src/generate_sample_data.py [hunk 1]
```diff
@@ -287,8 +382,42 @@ def _build_matrix_export(base: list[dict[str, object]], rng: random.Random) -> p
...
+    plan_ids = sorted({*INHERITED_PLAN_IDS, "400001ABC", "300005R"})
+    reserved_ssns = {"111223333", "222334444", "333445555", "444556666", "555667777"}
+    used_ssns = set(reserved_ssns)
+
+    rows: list[dict[str, object]] = []
+    for _ in range(100):
+        ssn = f"{rng.randint(100000000, 999999999)}"
+        while ssn in used_ssns:
+            ssn = f"{rng.randint(100000000, 999999999)}"
+        used_ssns.add(ssn)
```
[the remark this answers is not in the document]
Comment (manuel-reyes-ml) [reply src/generate_sample_data.py]: Very good point. Yes the main purpose of this issue is to generate fake data using Faker() Python module without affecting current engine and outputs generators. Next issue to solve, is to make faker to reproduce edge-cases to test pipelines completely/.
